lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -266,6 +268,12 @@ class TypeRefinementContextBuilder : private ASTWalker {
       return std::make_pair(false, S);
     }
 
+    if (auto *CS = dyn_cast<CaseStmt>(S)) {
+      // HACK: Refinement context will be built after type checking case labels
+      // via direct call to buildCaseStmtRefinementContext
+      return std::make_pair(false, S);
```

Comment:
There's got to be a way to do this independent of the ordering of TypeCheckStmt. This should be modeled after the other statement walkers. When you find a `SwitchStmt`, you'll traverse down each of its component cases and expand a context for each one.

Reply:
The problem here is that resolving type of pattern in switch depends on name lookup which in turn requires TypeRefinementContext, so we need somehow break that cycle. This is described in more detail [here](https://forums.swift.org/t/how-to-typecheck-pattern-in-casestmt-before-building-typerefinementcontext/32284)

Reply:
Yuck.  Then maybe the modeling for this needs to change before we can take this. Otherwise we'll potentially receive conflicting answers to availability questions depending on how far along type checking is, which is less than ideal.